pkg/controller/installation/products/ups/reconciler_test.go

```diff
@@ -0,0 +1,221 @@
+package ups
+
+import (
+	"context"
+	"testing"
+
+	upsv1alpha1 "github.com/aerogear/unifiedpush-operator/pkg/apis/push/v1alpha1"
+	v1Route "github.com/openshift/api/route/v1"
+
+	"github.com/integr8ly/integreatly-operator/pkg/apis/integreatly/v1alpha1"
+	"github.com/integr8ly/integreatly-operator/pkg/controller/installation/marketplace"
+	"github.com/integr8ly/integreatly-operator/pkg/controller/installation/products/config"
+
+	moqclient "github.com/integr8ly/integreatly-operator/pkg/client"
+	"github.com/pkg/errors"
+	k8serr "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+func basicConfigMock() *config.ConfigReadWriterMock {
+	return &config.ConfigReadWriterMock{
+		ReadUpsFunc: func() (ups *config.Ups, e error) {
+			return config.NewUps(config.ProductConfig{}), nil
+		},
+		WriteConfigFunc: func(config config.ConfigReadable) error {
+			return nil
+		},
+	}
+}
+
+func errorConfigMock() *config.ConfigReadWriterMock {
+	return &config.ConfigReadWriterMock{
+		ReadUpsFunc: func() (ups *config.Ups, e error) {
+			return config.NewUps(config.ProductConfig{}), nil
+		},
+		WriteConfigFunc: func(config config.ConfigReadable) error {
+			return errors.New("some error")
+		},
+	}
+}
+
+func basicRouteMock() *v1Route.Route {
+	return &v1Route.Route{
+		TypeMeta: v1.TypeMeta{
+			Kind:       "Route",
+			APIVersion: "v1",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Namespace: "ups",
+			Name:      defaultRoutename,
+		},
+	}
+}
+
+func TestReconciler_ReconcileCustomResource(t *testing.T) {
+	scheme := runtime.NewScheme()
+	upsv1alpha1.SchemeBuilder.AddToScheme(scheme)
+
+	cases := []struct {
+		Name           string
+		client         client.Client
+		FakeConfig     *config.ConfigReadWriterMock
+		Installation   *v1alpha1.Installation
+		ExpectErr      bool
+		ExpectedStatus v1alpha1.StatusPhase
+		FakeMPM        *marketplace.MarketplaceInterfaceMock
+	}{
+		{
+			Name:           "UPS Test: test custom resource is reconciled and phase complete returned",
+			ExpectedStatus: v1alpha1.PhaseCompleted,
+			FakeMPM:        &marketplace.MarketplaceInterfaceMock{},
+			Installation:   &v1alpha1.Installation{},
+			FakeConfig:     basicConfigMock(),
+			client: fake.NewFakeClientWithScheme(scheme,
+				&upsv1alpha1.UnifiedPushServer{
```

Comment:
Since this `ups` object is being used multiple times it would be nice to pull it out into a separate function, passing in the status as an argument?